### ISSUE TITLE
Will support IPv6 link-local dns server address?

### ISSUE THREAD

**Issue by lianhansheng** [c-ares/c-ares#462]
Dns server address like  fe80::cd6:1e60:fe42:b8b4%wlan0

**Comment by bradh352** [comment]
@sergvpurik since you provided the original PR, can you see if PR #646 does what is needed?

@lianhansheng not sure if you can test this

**Comment by sergvpurik** [comment]
@bradh352 
> can you see if PR https://github.com/c-ares/c-ares/pull/646 does what is needed?

Yes, it works.


For testing, I used IPv6 link-local IP of my router.
I use OpenWrt on my route, so it is easy to get console and find IPs.

More detailed steps:
1. Create "/etc/test-resolv.conf" with a router IP and specify a local interface name
```
echo nameserver fe80::xxxx:xxxx:xxxx:ae3e%wlx04xxxx > /etc/test-resolv.conf
```
2. Point system resolv.conf to /etc/test-resolv.conf
```
ln -s /etc/test-resolv.conf /etc/resolv.conf
```
3. Use c-ares adig tool to check DNS resolving
```
./adig -d www.facebook.com
```

4.  Restore default resolv.conf
```
cd /etc
ln -fsv ../run/systemd/resolve/stub-resolv.conf ./resolv.conf
```

Also, MacBook internet sharing can return IPv6 link-local for DNS and gateway.
Maybe, as I remember, the NAT64 option should be enabled.

**Comment by bradh352** [comment]
You mentioned Mac here, have you actually been able to get your mac's dns configuration to show a link-local ipv6 address?
I have a MacOS vm and an Ubuntu VM both bridged to the same network adapter.  My Ubuntu VM picks up the router advertisement for DNS, but my Mac VM doesn't appear to at least according to:

`scutil --dns`

That said, I'm also getting an ipv4 dns server from DHCP so I don't know if that's causing the link-local RA to be ignored.

I have a similar issue on Windows, I do see it picking up the RA for the ipv6 default route, but not the RDNSS record.

Both Windows and MacOS work if I specify the link-local server to 'adig' with the appropriate interface, just not getting the system to actually record the dns server to properly test that aspect.

**Comment by sergvpurik** [comment]
I don't have Mac for now. But I have an old log where I used rdisc6 and rdnssd.
I've tested this case on Linux, so I can say nothing about Mac or Windows clients.

The actual configuration of my Linux client was the next:
```
IP:       2001:2:0:xxxx:xxxx:xxxx:xxxx:b9
Gateway:  fe80::xxxx:xxxx:xxxx:5764
DNS:      fe80::xxxx:xxxx:xxxx:5764%wlan0
```
As the interface IP was IPv6, I suppose NAT64 option was used (Mac Internet sharing).

**Comment by bradh352** [comment]
fixed with #646
